lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs

```diff
@@ -0,0 +1,1015 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::errors::invalid_data_error;
+use crate::{vchan, Payload};
+use bitflags::bitflags;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use num_traits::FromPrimitive;
+use rdp::core::{mcs, tpkt};
+use rdp::model::data::Message;
+use rdp::model::error::*;
+use rdp::try_let;
+use std::collections::HashMap;
+use std::io::{Read, Write};
+use std::sync::Mutex;
+
+pub const CHANNEL_NAME: &str = "cliprdr";
+
+/// Client implements a client for the clipboard virtual channel
+/// (CLIPRDR) extension, as defined in:
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/fb9b7e0b-6db4-41c2-b83c-f889c1ee7688
+pub struct Client {
+    clipboard: Mutex<HashMap<u32, Vec<u8>>>,
```

Comment:
I may be missing something, but does this actually need to be under a mutex? It doesn't seem like we need to update this map anywhere we only have a shared reference.

Reply:
I was being cautious here but I think you're right, since all accesses are behind the mutex for `RdpClient`.